oshdb/src/test/java/org/heigit/ohsome/oshdb/contribution/ContributionsRelationTest.java

```diff
@@ -0,0 +1,315 @@
+
+    var osh = osh(versions, emptyList(), emptyList());
+    var contribs = Contributions.of(osh);
+
+    assertTrue(contribs.hasNext());
+    var contrib = contribs.next();
+    assertContrib(1000, 101, 1, versions.get(0), contrib);
+    assertEquals(EnumSet.of(CREATION), contrib.getTypes());
+
+    assertFalse(contribs.hasNext());
+  }
+
+  @Test
+  void testMissingMemberChanges() {
+    var nodes = List.of(
+        osh(1,
+            node(1, 1000, 101, 1, tags(), 0, 0)));
+    var ways = List.of(
+        osh(1, List.of(
+            osh(1,
+                node(1, 1000, 101, 1, tags(), 0, 0))),
+            way(1, 1000, 101, 1, tags(), mems(1))));
+
+    var versions = relations(1,
+        relation(1, 1000, 101, 1, tags(), mems(w(1, 0), r(1, 0))));
+
+    var osh = osh(versions, nodes, ways);
+    var contribs = Contributions.of(osh);
+
+    assertTrue(contribs.hasNext());
+    var contrib = contribs.next();
+    assertContrib(1000, 101, 1, versions.get(0), contrib);
+    assertEquals(EnumSet.of(CREATION), contrib.getTypes());
+
+    assertFalse(contribs.hasNext());
+  }
+
+  @Test
+  void testDifferentWayAndNode() {
+    var nodes = List.of(
+        osh(1,
+            node(1, 1000, 101, 1, tags(), 0, 0)),
+        osh(2,
+            node(1, 1000, 101, 1, tags(), 0, 0)),
+        osh(3,
+            node(1, 1000, 101, 1, tags(), 0, 0)));
+    var ways = List.of(
+        osh(1, List.of(
+            osh(1,
+                node(1, 1000, 101, 1, tags(), 0, 0))),
+            way(1, 1000, 101, 1, tags(), mems(1))),
+        osh(2, List.of(
+            osh(1,
+                node(1, 1000, 101, 1, tags(), 0, 0))),
+            way(1, 1000, 101, 1, tags(), mems(1))),
+        osh(3, List.of(
+            osh(1,
+                node(1, 1000, 101, 1, tags(), 0, 0))),
+            way(1, 1000, 101, 1, tags(), mems(1))));
+
+    var versions = relations(1,
+        relation(2, 2000, 202, 2, tags(),
+            mems(w(1, 0), n(1, 0), w(2, 0), n(2, 0), w(1, 0), n(1, 0))),
+        relation(1, 1000, 101, 1, tags(), mems(w(1, 0), n(1, 0), w(3, 0), n(3, 0))));
+
+    var osh = osh(versions, nodes, ways);
+    var contribs = Contributions.of(osh);
+
+    assertTrue(contribs.hasNext());
+    var contrib = contribs.next();
+    assertContrib(2000, 202, 2, versions.get(0), contrib);
+    assertEquals(EnumSet.of(MEMBER_CHANGE), contrib.getTypes());
+
+    assertTrue(contribs.hasNext());
+    contrib = contribs.next();
+    assertContrib(1000, 101, 1, versions.get(1), contrib);
+    assertEquals(EnumSet.of(CREATION), contrib.getTypes());
+
+    assertFalse(contribs.hasNext());
+  }
+
+  @Test
+  void testEmptyMemberContribs() {
+    var nodes = List.of(
+        osh(1,
+            node(1, 3000, 303, 3, tags(), 0, 0)),
+        osh(2,
+            node(1, 2000, 202, 2, tags(), 0, 0)));
+
+    var versions = relations(1,
+        relation(3, 3000, 303, 3, tags(), mems(1, 2)),
+        relation(2, 2000, 202, 2, tags(), mems(2)),
```

Comment:
Would it be worth to have (either in another or this) test that checks if at timestamp `2000` only node 2 exists as member but you still use `mems(1, 2)` in this line?

Reply:
Actually timestamp `2000` is not the "test or problem" here.
relation version 1 timestamp `1000` reference to a member which didn't not exist for this timestamp.